Refactored math tests to iterate over all math ops

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21512 Added math.log2 and hypot
* **#21511 Refactored math tests to iterate over all math ops**

Added similar logic to __init__.py

Formatting

Added compatibility for python2

Differential Revision: [D16088193](https://our.internmc.facebook.com/intern/diff/D16088193)

Chillee committed:
Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

ailzhang commented on test/test_jit.py:
```diff
@@ -6137,33 +6138,53 @@ def func(a, b):
+
+        unimplemented = ["fsum", "isclose", "trunc", "hypot", "log2"]
+        ops = [x for x in dir(math) if callable(getattr(math, x))]
+        for op in ops:
```
Hmmm I somehow feel it's cleaner as old way than this (growing) giant if-else block. cc: @wanchaol @eellison What do you think?

Chillee replied:
well, really, we needed to have a bunch of special cases for calling the different tests anyways. So for each special case for calling the tests, we have an additional if branch, so I don't think it's too bad.

eellison replied:
I think it's going to be messy no matter what. I guess my general preference is to avoid relying on python implementation details when possible, which we're doing here by calling `dir`. That has been safe so far but who knows what future python versions will do.  (Although we do that in other places too @driazati)

In this case i guess slight preference towards not relying on `dir`, but it's test code and kind of arbitrary and messy so i think it's up to the author here.

Chillee replied:
I think the upside of not needing to manually compare each function in the math documentation 
 with our implementations to see what's implemented is worth the downside of relying on python implementation details. For example, the original issue missed the functions `log2` and `hypot`, presumably because they added them in largely manually.

I also don't think `dir` is really a "python implementation detail". We use it all over the place, and not just in tests.

eellison replied:
`dir` is picking up attributes that are not public facing methods of the module such as `__loader__`. That feels like implementation details to me. Where else do we use `dir`? I know we do in some places where there isn't an alternative

driazati replied:
`dir` doesn't get implementation details, everything it gets is well-defined. Ignoring `__whatever__` will leave us with just the public `math` API, which is what we want

eellison replied:
```
Because dir() is supplied primarily as a convenience for use at an interactive prompt, it tries to supply an interesting set of names more than it tries to supply a rigorously or consistently defined set of names, and its detailed behavior may change across releases. For example, metaclass attributes are not in the result list when the argument is a class.
```

It's not a stable API

ailzhang reviewed (commented):
Refactor in `__init__.py` looks great, my only concern is about the if-else block in test.

eellison commented on test/test_jit.py:
```diff
@@ -6137,33 +6138,53 @@ def func(a, b):
+        for op in ops:
+            if op in unimplemented:
+                continue
+            elif op == "__loader__":    # Weird windows behavior...
```
Can you put __loader__ in unimplemented (maybe with comment) ?

Chillee replied:
I think I'd prefer to have it separately, as it's not a math function. On my computer, it's not even a function...

eellison replied:
word, this is the downside of this approach, is you get random python implementation details

eellison commented on torch/jit/__init__.py:
```diff
@@ -1978,9 +1935,15 @@ def register_all(mod):
+    unimplemented_math_ops = ["fsum", "isclose", "trunc", "hypot", "log2"]
+    special_math_ops = ["remainder"]    # We bound this to aten::mathremainder.
+    for op in math_ops:
+        if op == "__loader__":  # Weird windows behavior...
```
reason not to put op in unimplemented_math_ops ?

Chillee replied:
Same reasoning as above I think.

driazati commented on test/test_jit.py:
```diff
@@ -6137,33 +6138,53 @@ def func(a, b):
+                        prec = (1e-4)
+                    self.assertEqual(res_python, res_script, message=msg, prec=prec)
+
+        unimplemented = ["fsum", "isclose", "trunc", "hypot", "log2"]
```
This should pull its list from `torch/jit/__init__.py` so they don't get out of sync

driazati commented on torch/jit/__init__.py:
```diff
@@ -1978,9 +1935,15 @@ def register_all(mod):
+    unimplemented_math_ops = ["fsum", "isclose", "trunc", "hypot", "log2"]
+    special_math_ops = ["remainder"]    # We bound this to aten::mathremainder.
+    for op in math_ops:
+        if op == "__loader__":  # Weird windows behavior...
```
Why don't we just ignore all things that start with `__`?

Chillee replied:
Yeah that'd probably be better.

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head

Chillee committed:
Update on "Refactored math tests to iterate over all math ops"

Refactored math tests to iterate over all math ops

Added similar logic to __init__.py

Formatting

Added compatibility for python2

gh-metadata: pytorch pytorch 21511 gh/chillee/35/head
